beluga_amcl/include/beluga_amcl/filter_update_control/filter_update_control_mixin.hpp

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_AMCL__FILTER_UPDATE_CONTROL__FILTER_UPDATE_CONTROL_MIXIN_HPP_
+#define BELUGA_AMCL__FILTER_UPDATE_CONTROL__FILTER_UPDATE_CONTROL_MIXIN_HPP_
+
+#include <utility>
+#include <vector>
+#include <execution>
+
+#include <beluga/sensor.hpp>
+#include <beluga_amcl/tf2_sophus.hpp>
+
+#include <ciabatta/ciabatta.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a resampling policies poller.
+ */
+
+namespace beluga_amcl
+{
+
+/// @brief Public interface of the FilterUpdateControl filter customization mixin
+class FilterUpdateControlInterface
+{
+public:
+  using motion_update_type = Sophus::SE2d;
+  using sensor_update_type = std::vector<std::pair<double, double>>;
+
+  /// @brief Default destructor
+  virtual ~FilterUpdateControlInterface() = default;
+
+  /** @brief Execute a filter update step, conditioned by the installed resampling policies.
+   *  @param odom_to_base_transform Odom-to-base transform, for the motion update.
+   *  @param laser_scan Laser-scan information, for the sensor update.
+   *  @return true if the estimation result of the filter has been updated, false otherwise.
+   */
+  virtual bool update_filter(
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+
+  /** \overload
+   * It allows specifying a sequenced execution policy.
+   */
+  virtual bool update_filter(
+    std::execution::sequenced_policy exec_policy,
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+
+  /** \overload
+   * It allows specifying a parallel execution policy.
+   */
+  virtual bool update_filter(
+    std::execution::parallel_policy exec_policy,
+    const motion_update_type & odom_to_base_transform,
+    sensor_update_type && laser_scan) = 0;
+};
+
+/// Resampling policy poller.
+/**
+ * \tparam Mixin The mixed-in type.
+ * \tparam UpdateFilterWhenMovingPolicy The resample on motion policy type.
+ * \tparam ResampleIntervalPolicy The resample interval policy type.
+ * \tparam SelectiveResamplingPolicy The selective resampling policy type.
+ */
+template<
+  typename Mixin,
+  typename UpdateFilterWhenMovingPolicy,
+  typename ResampleIntervalPolicy,
+  typename SelectiveResamplingPolicy>
```

Comment:
They are, and two of them could be injected using Dynamic Dependency Injection, but the SelectiveResample one needs access to the filter instance. I kept them all as templates only not to have two different types of insertion.

Reply:
Right, I mean, they are so specific that I'm questioning if it even makes sense to inject them in any way.

Why aren't the members of the `FilterUpdateControlMixin` class concrete instances of `ResampleIntervalPolicy`, `UpdateFilterWhenMovingPolicy` and `SelectiveResamplingPolicy` instead of being templates?

Reply:
Hmmm, I see. We can do that. The one downside is that testing FilterUpdateControlMixin will be slightly more complex, since mocking will be off the table.

Reply:
Fair enough.